client/rest/src/main/java/org/elasticsearch/client/RequestOptions.java

```diff
@@ -0,0 +1,153 @@
+        return builder;
+    }
+
+    private static final Header[] NO_HEADERS = new Header[0];
+
+    private final Header[] headers;
+    private final HttpAsyncResponseConsumerFactory httpAsyncResponseConsumerFactory;
+
+    private RequestOptions(Builder builder) {
+        this.headers = builder.headers;
+        this.httpAsyncResponseConsumerFactory = builder.httpAsyncResponseConsumerFactory;
+    }
+
+    public Builder toBuilder() {
+        Builder builder = new Builder();
+        builder.setHeaders(headers);
+        builder.setHttpAsyncResponseConsumerFactory(httpAsyncResponseConsumerFactory);
+        return builder;
+    }
+
+    /**
+     * Headers to attach to the request.
+     */
+    public Header[] getHeaders() {
+        return headers;
+    }
+
+    /**
+     * The {@link HttpAsyncResponseConsumerFactory} used to create one
+     * {@link HttpAsyncResponseConsumer} callback per retry. Controls how the
+     * response body gets streamed from a non-blocking HTTP connection on the
+     * client side.
+     */
+    public HttpAsyncResponseConsumerFactory getHttpAsyncResponseConsumerFactory() {
+        return httpAsyncResponseConsumerFactory;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder b = new StringBuilder();
+        b.append("RequestOptions{");
+        if (headers.length > 0) {
+            b.append(", headers=");
+            for (int h = 0; h < headers.length; h++) {
+                if (h != 0) {
+                    b.append(',');
+                }
+                b.append(headers[h].toString());
+            }
+        }
+        if (httpAsyncResponseConsumerFactory != HttpAsyncResponseConsumerFactory.DEFAULT) {
+            b.append(", consumerFactory=").append(httpAsyncResponseConsumerFactory);
+        }
+        return b.append('}').toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || (obj.getClass() != getClass())) {
+            return false;
+        }
+        if (obj == this) {
+            return true;
+        }
+
+        RequestOptions other = (RequestOptions) obj;
+        return Arrays.equals(headers, other.headers)
+                && httpAsyncResponseConsumerFactory.equals(other.httpAsyncResponseConsumerFactory);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(Arrays.hashCode(headers), httpAsyncResponseConsumerFactory);
+    }
+
+    public static class Builder {
+        private Header[] headers;
+        private HttpAsyncResponseConsumerFactory httpAsyncResponseConsumerFactory;
+
+        /**
+         * Build the {@linkplain RequestOptions}.
+         */
+        public RequestOptions builder() {
+            return new RequestOptions(this);
+        }
+
+        /**
+         * Set the headers to attach to the request.
+         */
+        public void setHeaders(Header... headers) {
```

Comment:
Since this is a builder, I do think it'd be nice to be able to add headers iteratively, with something like

```java
builder.addHeader("X-Header", "foo");
builder.addHeader("User-Agent", "bob");
```

Reply:
I'd say that's the plan given that we just made the same change to the existing Request.